### DISCUSSION TITLE
How to use database-level default values for attributes?

### DISCUSSION THREAD

**Discussion by bfontaine** [fastapi-users/fastapi-users#668] [Q&A] [answered]
Hello,
I’m trying to use non-nullable default values for attributes at the database level, but I can’t get them to work. Starting from the [official SQLAlchemy example](https://frankie567.github.io/fastapi-users/configuration/full-example.html#sqlalchemy), I add a `foo` attribute to the user:

```diff
@@ -5,13 +5,15 @@ from fastapi_users import FastAPIUsers, models
 from fastapi_users.authentication import JWTAuthentication
 from fastapi_users.db import SQLAlchemyBaseUserTable, SQLAlchemyUserDatabase
 from sqlalchemy.ext.declarative import DeclarativeMeta, declarative_base
+from typing import Optional
+import sqlalchemy as sa
 
 DATABASE_URL = "sqlite:///./test.db"
 SECRET = "SECRET"
 
 
 class User(models.BaseUser):
-    pass
+    foo: str
 
 
 class UserCreate(models.BaseUserCreate):
@@ -31,7 +32,7 @@ Base: DeclarativeMeta = declarative_base()
 
 
 class UserTable(Base, SQLAlchemyBaseUserTable):
-    pass
+    foo = sa.Column(sa.String, nullable=False, server_default="foobar")
```

This doesn’t when I register a user because that field is required:
```
pydantic.error_wrappers.ValidationError: 1 validation error for UserDB
foo
  field required (type=value_error.missing)
```

Okay, so what if I make it optional:
```diff
@@ -12,7 +12,7 @@ DATABASE_URL = "sqlite:///./test.db"
 SECRET = "SECRET"
 
 class User(models.BaseUser):
-    foo: str
+    foo: Optional[str] = None
 
 
 class UserCreate(models.BaseUserCreate):
```

It now fails because it can’t be `null`: `sqlite3.IntegrityError: NOT NULL constraint failed: user.foo`. Note it’s the same if I use `foo: Optional[str]` and `foo: Optional[str] = None`.

If I remove it from Pydantic I can register a user, but I don’t get the field in the response:
```diff
@@ -12,7 +12,7 @@ DATABASE_URL = "sqlite:///./test.db"
 SECRET = "SECRET"
 
 class User(models.BaseUser):
-    foo: str
+    pass
```
```javascript
{"id":"86368e73-41a3-49b9-978c-d0a3139b731b","email":"9165@example.com","is_active":true,"is_superuser":false,"is_verified":false}
```

Notes:
- I have the same issue using `default=` and `server_default=`
- I don’t want to duplicate the default value in the Pydantic model

What am I doing wrong? Thanks for your help.

**Comment by frankie567** [comment, marked as answer]
Hi there,

Sorry for the delayed response 🙂

Unfortunately, the way SQLAlchemy adapter is implemented right now doesn't allow to have SQL defaults. There'll likely be work on a new adapter for SQLAlchemy 1.4 that should allow that.

In the meantime, you should use a Pydantic default.

**Reply by bfontaine** [reply]
Thank you :+1:
